crates/stackable-certs/src/cert_builder.rs

```diff
@@ -0,0 +1,330 @@
+/// };
+///
+/// let ca: CertificateAuthority<ecdsa::SigningKey> = CertificateAuthorityBuilder::builder()
+///     .build()
+///     .build_ca()
+///     .expect("failed to build CA");
+///
+/// let certificate = CertificateBuilder::builder()
+///     .subject("CN=trino-coordinator-default-0")
+///     .signed_by(&ca)
+///     .build()
+///     .build_certificate()
+///     .expect("failed to build certificate");
+/// ```
+#[derive(Builder)]
+pub struct CertificateBuilder<'a, KP>
+where
+    KP: CertificateKeypair,
+    <KP::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    /// Required subject of the certificate, usually starts with `CN=`.
+    subject: &'a str,
+
+    /// Optional list of subject alternative names (SAN) DNS entries,
+    /// that are added to the certificate.
+    #[builder(default)]
+    subject_alterative_dns_names: &'a [&'a str],
+
+    /// Validity/lifetime of the certificate.
+    ///
+    /// If not specified the default of [`DEFAULT_CERTIFICATE_VALIDITY`] will be used.
+    #[builder(default = DEFAULT_CERTIFICATE_VALIDITY)]
+    validity: Duration,
+
+    /// Serial number of the generated certificate.
+    ///
+    /// If not specified a random serial will be generated.
+    serial_number: Option<u64>,
+
+    /// Cryptographic keypair used to for the certificates.
+    ///
+    /// If not specified a random keypair will be generated.
+    key_pair: Option<KP>,
+
+    /// Mandatorily sign the certificate using the provided [`CertificateAuthority`].
+    signed_by: &'a CertificateAuthority<KP>,
+}
+
+impl<KP> CertificateBuilder<'_, KP>
+where
+    KP: CertificateKeypair,
+    <KP::SigningKey as signature::Keypair>::VerifyingKey: EncodePublicKey,
+{
+    pub fn build_certificate(
+        self,
+    ) -> Result<CertificatePair<KP>, CreateCertificateError<KP::Error>> {
+        let serial_number =
+            SerialNumber::from(self.serial_number.unwrap_or_else(|| rand::random::<u64>()));
+
+        let validity = Validity::from_now(*self.validity).context(ParseValiditySnafu)?;
+        let subject: Name = self.subject.parse().context(ParseSubjectSnafu {
+            subject: self.subject,
+        })?;
+        let key_pair = match self.key_pair {
+            Some(key_pair) => key_pair,
+            None => KP::new().context(CreateKeyPairSnafu)?,
+        };
+
+        let ca_validity = self.signed_by.ca_cert().tbs_certificate.validity;
+        let ca_not_after = ca_validity.not_after.to_system_time();
+        let cert_not_after = validity.not_after.to_system_time();
+        if ca_not_after < cert_not_after {
+            warn!(
+                ca.validity = ?ca_validity,
+                cert.validity = ?validity,
+                ca.not_after = ?ca_not_after,
+                cert.not_after = ?cert_not_after,
+                subject = ?subject,
+                "The lifetime of certificate authority is shorted than the lifetime of the generated certificate",
+            );
+        }
```

Comment:
praise: Thank you for adding this.

note: This should be a hard error instead.